_posts/2019-08-15-docker-containers-pt-1.markdown

```diff
@@ -0,0 +1,45 @@
+---
+layout: post
+title: "Roll your own Docker containers (part 1)"
+date: 2019-08-15 12:30:00
+categories: ["DevOps", "docker", "tutorials"]
+author: "rdormer"
+---
+
+In my [last article](https://www.ombulabs.com/blog/devops/docker/tutorials/gentle-intro-to-docker.html), I gave a brief tour of what Docker is, and basic usage. Once you've got your feet under you with basic usage, you'll reach the point where you want to start making your own containers. You'll certainly need to know how to roll your own if you want to use Docker in your own infrastructure.
+
+## Container basics
+
+Broadly speaking, there are two types of Docker images - *base images*, and images that are built on top of them. Most of the time, when you're making an image of your own, you'll be inheriting from a base (or parent) image, and tweaking it by layering changes over top of it.  Layering is exactly what you're dong, by the way - as you execute commands and copy data into a container, all of the changes you're making are written to a *writable layer* (also known as the *container layer*). There's a lot more to this particular subject - see [storage drivers](https://docs.docker.com/storage/storagedriver/) and [volumes](https://docs.docker.com/storage/volumes/) for a lot more information.
+
+## Starting the easy way
+
+You'll start by creating a Dockerfile, the configuration file containing the commands that Docker will use to build your image.  [The first command](https://docs.docker.com/engine/reference/builder/#from) is `From`:
+
+`FROM debian`
+
+Which specifies that you'll be building your image on top of [Debian](https://hub.docker.com/_/debian).  Usually, the image that you'll be building will be, at least partially, a copy of some application that you're running.  You can create an image of this application by putting your Dockerfile in a directory containing a clean installation of your application. All of the commands in your Dockerfile will be relative to this directory, which Docker calls the *context*.  Add the [following command](https://docs.docker.com/engine/reference/builder/#copy) after the `FROM` command:
+
+`COPY . /myapp`
+
+If you have [any other commands that need to be run](https://docs.docker.com/engine/reference/builder/#run), you can use the `CMD` command.  Let's use it to install [RVM](https://rvm.io/):
+
+`CMD apt-get update`
```

Comment:
`CMD` is not used like that.

You would like to replace that with `RUN`

See: https://goinbigdata.com/docker-run-vs-cmd-vs-entrypoint/